### PR TITLE
Refactor faceconverter

### PR DESCRIPTION
**Pull request by ghost**
Request for comment on basic refactoring of FaceConverter.h

- The only conversion function fully implemented is convertIfcPlane. The remaining partial implementations were removed for now to get an overview of FaceConverter.h. The idea is to refactor each function body, add it back into the function commented and make the full implementation of each function an issue on Github. 

- Would it also be possible to implement the parameters "pos" and "polylineData" at the lowest level function (e.g. convertIfcBSplineSurface)? To avoid handing them down from e.g.  convertIfcSurface to convertIfcBoundedSurface to convertIfcBSplineSurface?

### CODE REVIEW

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 1]
```diff
@@ -40,13 +40,25 @@ namespace OpenInfraPlatform {
...
 		namespace IfcGeometryConverter {
 			class ItemData;
-
+			
+			/*! \brief Converter functionality for \c IfcSurfaces's subtypes.
+            		*
+          		* This class includes converter functions for
+            		* \c IfcBoundedSurface and its subtypes,
+            		* \c IfcElementarySurface and its subtypes and
+           	 	* \c IfcSweptSurface.
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h]: I think there are also IfcFace converters. And I also added some from the representationConverter (see #147 ).

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 2]
```diff
@@ -22,922 +22,340 @@
...
-					std::shared_ptr<UnitConverter<IfcEntityTypesT>> unitConverter,
-					std::shared_ptr<PlacementConverterT<IfcEntityTypesT>> pc,
-					std::shared_ptr<CurveConverterT<IfcEntityTypesT>> cc)
-					: 
+
+				//! Constructor
+				FaceConverterT(EXPRESSReference<GeometrySettings> geomSettings,
+					EXPRESSReference<UnitConverter<IfcEntityTypesT>> unitConverter,
+					EXPRESSReference<PlacementConverterT<IfcEntityTypesT>> pc,
+					EXPRESSReference<CurveConverterT<IfcEntityTypesT>> cc)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:60]: These should stay std::shared_ptr.

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 3]
```diff
@@ -22,922 +22,340 @@
...
+					\param		pos
+					\param		polylineData
+					\return
+					\note		The \c IfcSurface subtypes are \c IfcBoundedSurface,  \c IfcElementarySurface and  \c IfcSweptSurface.
+				*/
 
-						IfcSurface
+				void convertIfcSurface(const EXPRESSReference<typename IfcEntityTypesT::IfcSurface>& surface,
+					const carve::math::Matrix& pos,
+					EXPRESSReference<carve::input::PolylineSetData>& polylineData) {
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:86]:
```suggestion
					std::shared_ptr<carve::input::PolylineSetData>& polylineData) 
				{
```
Carve stuff cannot be EXPRESSreference, only IFC stuff.

Comment (pjanck) [reply Core/src/IfcGeometryConverter/FaceConverter.h:86]: Also, make all functions `const throw(...)`
```suggestion
					std::shared_ptr<carve::input::PolylineSetData>& polylineData) const throw(...)
				{
```

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 4]
```diff
@@ -22,922 +22,340 @@
...
 					*/
 
-					// ************************************************************************************************************************	//
-					//	IfcBoundedSurface SUPTYPE of IfcSurface																					//
-					//	ABSTRACT SUPERTYPE of IfcBSplineSurface, IfcCurveBoundedPlane, IfcCurveBoundedSurface, IfcRectangularTrimmedSurface		//
-					// ************************************************************************************************************************	//
-
-					std::shared_ptr<typename IfcEntityTypesT::IfcBoundedSurface> bounded_surface = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcBoundedSurface>(surface);
-
-					if(bounded_surface) {
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:103]: None of the functionalities can be lost. Pack it in another function, yes, but do not delete code that is there.

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 5]
```diff
@@ -22,922 +22,340 @@
...
-						throw oip::UnhandledException(surface);
+					// (4/4) IfcRectangularTrimmedSurface SUBTYPE of IfcBoundedSurface.
+					else if (surface.isOfType<typename IfcEntityTypesT::IfcRectangularTrimmedSurface>()) {
+						return convertIfcRectangularTrimmedSurface(surface.as<typename IfcEntityTypesT::IfcRectangularTrimmedSurface>(), pos, polylineData);
 					}
-					throw oip::UnhandledException(surface);
 				}
 
-				// TODO: What is happening here?
-				void convertIfcFaceList(const std::vector<std::shared_ptr<typename IfcEntityTypesT::IfcFace>>& faces,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:408]: Where is this function now?

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 6]
```diff
@@ -22,922 +22,340 @@
...
 
-					// IfcFaceList can be a closed or open shell,
-					// so let the calling function decide where to put it
-					item_data->open_or_closed_polyhedrons.push_back(polygon);
-				}
-
-				bool convertIfcFace(const std::shared_ptr<typename IfcEntityTypesT::IfcFace>& face,
-					const carve::math::Matrix& pos,
-					std::shared_ptr<carve::input::PolyhedronData> polygon,
-					std::map<std::string, uint32_t>& polygonIndices)
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:437]: Where is this function now?

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 7]
```diff
@@ -22,922 +22,340 @@
...
+					else if (surface.isOfType<typename IfcEntityTypesT::IfcSurfaceOfRevolution>()) {
+						return convertIfcSurfaceOfRevolution(surface.as<typename IfcEntityTypesT::IfcSurfaceOfRevolution>(), pos, polylineData);
 					}
-
-					triangulateFace(mergedVertices2D, mergedVertices3D, faceLoopReversed, polygon, polygonIndices);
-
-					return !convertionFailed;
 				}
 
-				bool convert3DPointsTo2D(const int boundID,
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:635]: This function is needed elsewhere - does you code compile successfully?

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 8]
```diff
@@ -22,922 +22,340 @@
...
+					\param	pos
+					\param	polylineData
+					\return
+					\note	The \c IfcCurveBoundedPlane is a subtype of \c IfcBoundedSurface.
+				*/
 
-					return true;
+				void convertIfcCurveBoundedPlane(const EXPRESSReference<typename IfcEntityTypesT::IfcCurveBoundedPlane>& surface,
+					const carve::math::Matrix& pos,
+					EXPRESSReference<carve::input::PolylineSetData>& polylineData) {
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h:233]:
```suggestion
					std::shared_ptr<carve::input::PolylineSetData>& polylineData
				) const throw(...)
				{
```
It's easier to read with a few more line breaks - please apply to all functions.

**Review by pjanck — CHANGES_REQUESTED** [review]
A lot of functionalities were lost, as code was deleted and not moved.

**Comment by pjanck** [comment]
Regarding your questions:
- It's easier for the git history, if code is moved immediately, and not deleted and then added at a later stage. Please see dc5d4f7 for an example.
- pos & polydata need to be handed from function to function. I can't imagine how to do it differently but am open to suggestions.

**Comment by hechth** [comment]
There seem to be a lot of old commits - please commit only changes which are related to the issue.

**Pull request closed without merging**